url4fdf79ae2c_beckchr_juel_tgz-pJ8-de_odysseus_el_samples_xml_sax_XMELFilterJ8/src/de/odysseus/el/tree/NodePrinter.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2006-2009 Odysseus Software GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package de.odysseus.el.tree;
+
+import java.io.PrintWriter;
+import java.util.Stack;
+
+/**
+ * Node pretty printer for debugging purposes.
+ *
+ * @author Christoph Beck
+ */
+public class NodePrinter {
+	private static boolean isLastSibling(Node node, Node parent) {
+		if (parent != null) {
+			return node == parent.getChild(parent.getCardinality() - 1);
+		}
+		return true;
+	}
+	
+	private static void dump(PrintWriter writer, Node node, Stack<Node> predecessors) {
+		if (!predecessors.isEmpty()) {
+			Node parent = null;
+			for (Node predecessor: predecessors) {
+				if (isLastSibling(predecessor, parent)) {
+					writer.print("   ");
+				} else {
+					writer.print("|  ");
+				}
+				parent = predecessor;
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 37:44, 48:55.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.